### PR TITLE
[C++ API] Add functional overloads for pixelshuffle, pooling, upsampling, vision

### PR DESCRIPTION
**Pull request by yf225**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29360 Add functional overloads for fold, linear, loss, normalization, padding
* **#29359 Add functional overloads for pixelshuffle, pooling, upsampling, vision**
* #29358 Add functional overloads for activation, batchnorm, distance, embedding

This PR adds functional overloads that take the full set of arguments (instead of just Options) for the following functionals:
- pixelshuffle
- pooling
- upsampling
- vision

These new functionals lives in the `torch::nn::functional::detail` namespace and they are only meant to be called from the module forward methods (i.e. they are not public API). This is in preparation for the future change where we make module Options and functional Options two different classes, because if the module forward method has to construct a new functional Options object every time it runs it will be pretty silly and bad performance.

Differential Revision: [D18376978](https://our.internmc.facebook.com/intern/diff/D18376978)

### CODE REVIEW

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit e12d49a:
* 0/1 recognized as flaky
* 1/1 failures introduced in this PR



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=e12d49a6d75f6a102d8c49302d4fc5a683b2c37c).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

File: torch/csrc/api/include/torch/nn/functional/vision.h [hunk 1]
```diff
@@ -47,52 +47,70 @@ inline Tensor affine_grid(
...
 }
 
+// ============================================================================
+
+namespace detail {
 inline Tensor grid_sample(
     const Tensor& input,
     const Tensor& grid,
-    GridSampleOptions options = {}) {
+    std::string mode = "bilinear",
```
Comment (pbelevich) [inline torch/csrc/api/include/torch/nn/functional/vision.h]: Are we going to change it to torch:: enums?

File: torch/csrc/api/include/torch/nn/functional/vision.h [hunk 2]
```diff
@@ -47,52 +47,70 @@ inline Tensor affine_grid(
...
-  return torch::grid_sampler(input, grid, mode_enum, padding_mode_enum, options.align_corners().value());
+
+  return torch::grid_sampler(input, grid, mode_enum, padding_mode_enum, align_corners.value());
+}
+} // namespace detail
+
+inline Tensor grid_sample(
+    const Tensor& input,
+    const Tensor& grid,
+    GridSampleOptions options = {}) {
```
Comment (pbelevich) [inline torch/csrc/api/include/torch/nn/functional/vision.h]: why by value?

**Comment by pbelevich** [comment]
If `torch::nn::functional::detail` methods are supposed to be called by forward methods and are not part of public API why do they have default values?

File: torch/csrc/api/include/torch/nn/functional/upsampling.h [hunk 3]
```diff
@@ -42,65 +48,75 @@ inline Tensor interpolate(const Tensor& input, InterpolateOptions options) {
...
         false,
         "Input Error: Only 3D, 4D and 5D input Tensors supported "
         "(got ", input.dim(), "D) for the modes: nearest | linear | bilinear | bicubic | trilinear "
-        "(got ", enumtype::get_enum_name(options.mode()), ")");
+        "(got ", enumtype::get_enum_name(mode), ")");
   }
 }
+} // namespace detail
+
+inline Tensor interpolate(const Tensor& input, InterpolateOptions options = {}) {
```
[the remark this answers is not in the document]
Comment (yf225) [reply torch/csrc/api/include/torch/nn/functional/upsampling.h]: In some of the functionals we want to be able to modify the fields in the options (such as `out` in `NormalizeFuncOptions` added in https://github.com/pytorch/pytorch/pull/29404), and Ed also suggested that we should just pass options by value because it's a small struct (https://github.com/pytorch/pytorch/pull/27942#discussion_r337604093).

File: torch/csrc/api/include/torch/nn/functional/upsampling.h [hunk 4]
```diff
@@ -9,30 +9,36 @@ namespace torch {
 namespace nn {
 namespace functional {
 
-inline Tensor interpolate(const Tensor& input, InterpolateOptions options) {
-  auto _check_size_scale_factor = [options](size_t dim) {
-    if (options.size().empty() && options.scale_factor().empty()) {
+namespace detail {
+inline Tensor interpolate(
+  const Tensor& input,
+  std::vector<int64_t> size = {},
```
[the remark this answers is not in the document]
Comment (yf225) [reply torch/csrc/api/include/torch/nn/functional/upsampling.h]: Fixed

File: torch/csrc/api/include/torch/nn/functional/vision.h (hunk 1 above)
Comment (yf225) [reply torch/csrc/api/include/torch/nn/functional/vision.h]: Yes I will do it in a follow-up PR (before the release)

File: torch/csrc/api/include/torch/nn/functional/vision.h (hunk 2 above)
Comment (yf225) [reply torch/csrc/api/include/torch/nn/functional/vision.h]: replied in an earlier comment

**Comment by yf225** [comment]
> If `torch::nn::functional::detail` methods are supposed to be called by forward methods and are not part of public API why do they have default values?

Thanks a lot for the suggestion - I removed the default values from all of them.

**Review by pbelevich — APPROVED** [review]
